### PR TITLE
Fix a bug for `_enforce_type` in pandas >= 1.0

### LINKED ISSUE #3006: [BUG] `_enforce_schema` raises `TypeError` in pandas >= 1.0.0 when a dataframe contains string column
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [x] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **MLflow installed from (source or binary)**:
- **MLflow version (run ``mlflow --version``)**:
- **Python version**:
- **npm version, if running the dev UI**:
- **Exact command to reproduce**:

### Describe the problem
Describe the problem clearly here. Include descriptions of the expected behavior and the actual behavior.

### Code to reproduce issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```python
import pandas as pd

import mlflow
from mlflow.pyfunc import _enforce_schema
from mlflow.types.utils import _infer_schema

print('mlflow version:', mlflow.__version__)
print('pandas version:', pd.__version__)

# works
df = pd.DataFrame({'num': [0.0]})
_enforce_schema(df, _infer_schema(df))

# raises an error
df = pd.DataFrame({'str': ['a']})
_enforce_schema(df, _infer_schema(df))
```

This outputs:

```
mlflow version: 1.9.1.dev0
pandas version: 1.0.0
Traceback (most recent call last):
  File "test.py", line 18, in <module>
    _enforce_schema(df, schema)
  File "/Users/xxx/Desktop/projects/mlflow/mlflow/pyfunc/__init__.py", line 372, in _enforce_schema
    new_pdf[x] = _enforce_type(x, pdf[x], col_types[i])
  File "/Users/xxx/Desktop/projects/mlflow/mlflow/pyfunc/__init__.py", line 288, in _enforce_type
    if values.dtype in (t.to_pandas(), t.to_numpy()):
TypeError: data type not understood
```

It seems the error occurs when comparing `np.dtype("O")` (`values.dtype`) and `pd.StringDtype()` (`t.to_pandas()`).

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

Fixes #3006 

## How is this patch tested?

Run the tests with `pandas>=1.0` and `pandas==0.23.4`.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: travis/large-requirements.txt [hunk 1]
```diff
@@ -11,7 +11,6 @@ onnxmltools==1.4.0;
 onnxruntime==0.3.0;
 mleap==0.16.0
 mxnet==1.5.0
-pandas<=0.23.4
```
Comment (harupy) [inline travis/large-requirements.txt:14]: Without this restriction, the latest version will be installed.

File: mlflow/pyfunc/__init__.py [hunk 2]
```diff
@@ -292,16 +292,6 @@ def _enforce_type(name, values: pandas.Series, t: DataType):
...
-        # The types are already compatible => conversion is not necessary.
-        return values
-
-    if t == DataType.binary and values.dtype.kind == t.binary.to_numpy().kind:
-        # NB: bytes in numpy have variable itemsize depending on the length of the longest
-        # element in the array (column). Since MLflow binary type is length agnostic, we ignore
-        # itemsize when matching binary columns.
-        return values
-
     if t == DataType.string and values.dtype == np.object:
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:295]: Just checking string first solved the issue.

File: mlflow/pyfunc/__init__.py [hunk 3]
```diff
@@ -292,16 +292,6 @@ def _enforce_type(name, values: pandas.Series, t: DataType):
...
         #  NB: strings are by default parsed and inferred as objects, but it is
         # recommended to use StringDtype extension type if available. See
```
Comment (harupy) [inline mlflow/pyfunc/__init__.py:297]: Reordering the checks solved the issue.

File: travis/large-requirements.txt (hunk 1 above)
Comment (harupy) [reply travis/large-requirements.txt:14]: https://github.com/mlflow/mlflow/pull/3130/checks?check_run_id=884982725#step:3:486

**Review by tomasatdatabricks — APPROVED** [review]
lgtm!

**Pull request merged**